acasclient/lsthing.py

```diff
@@ -971,24 +960,40 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
         :type code_kind: str
         :param code_origin: Origin of DDictValue referenced, typically 'ACAS DDict'
         :type code_origin: str
-        :param client: Authenticated acasclient.client instance to look up current DDictValues
-        :type client: acasclient.client
+        """
+        self.code = code
+        self.ddict = ddict
+        if ddict:
+            if isinstance(ddict, DDict):
+                self.code_type = ddict.code_type
+                self.code_kind = ddict.code_kind
+                self.code_origin = ddict.code_origin
+            else:
+                raise ValueError('ddict must be an instance of acasclient.ddict.DDict')
+        else:
+            self.code_type = code_type
+            self.code_kind = code_kind
+            self.code_origin = code_origin
+            # Auto-recognize some code origin:
+            if self.code_origin.upper() == ACAS_DDICT:
+                self.ddict = ACASDDict(code_type, code_kind)
+            elif self.code_origin.upper() == ACAS_LSTHING:
+                self.ddict = ACASLsThingDDict(code_type, code_kind)
+
+    def __hash__(self):
+        return hash(f'{self.code}-{self.code_type}-{self.code_kind}-'
+                    f'{self.code_origin}')
+
+    def validate(self):
+        """Validate that this CodeValue conforms to the stored list of possible DDictValues in self.ddict
+
         :return: Error message, or None if valid
         :rtype: str | None
```

Comment:
`ValueError` will be raised here in case of invalid ddict.

I believe we are going to remove the validation logic raising a `ValueError` exception, so this is optional based on what we end up doing.